mlflow/entities/trace_data.py

```diff
@@ -0,0 +1,22 @@
+from dataclasses import asdict, dataclass, field
+from typing import Any, Dict, List
+
+from mlflow.entities.span import Span
+
+
+@dataclass
+class TraceData:
+    """A container object that holds the spans data of a trace.
+
+    Args:
+        spans: List of spans that are part of the trace.
+    """
+
+    spans: List[Span] = field(default_factory=list)
+
+    @classmethod
+    def from_dict(cls, d):
+        return cls(spans=[Span(**span) for span in d["spans"]])
+
+    def to_dict(self) -> Dict[str, Any]:
+        return asdict(self)
```

Comment:
The problem I encountered is that asdict doesn't work well in json.dumps. For events, when we call add_event it invokes here https://github.com/open-telemetry/opentelemetry-python/blob/29c293fbb120ff8b760519a25e0530b5ab05af83/opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py#L836-L853 and the attributes field finally leads to include a lock here https://github.com/open-telemetry/opentelemetry-python/blob/29c293fbb120ff8b760519a25e0530b5ab05af83/opentelemetry-api/src/opentelemetry/attributes/__init__.py#L153 So my workaround is to dump attributes to string when saving the whole piece https://github.com/mlflow/mlflow/pull/11541/files#r1542348332

Reply:
The issue is addressed in this PR. The attribute field is converted to simple dict when creating `Span` object.